src/Monotonic.cpp

```diff
@@ -233,71 +385,58 @@ class MonotonicVisitor : public IRVisitor {
         op->b.accept(this);
-        Monotonic rb = result;
+        ConstantInterval rb = result;
         result = unify(ra, rb);
     }
 
     void visit(const Not *op) override {
         op->a.accept(this);
-        result = flip(result);
+        result = negate(result);
     }
 
     void visit(const Select *op) override {
-        op->condition.accept(this);
-        Monotonic rcond = result;
-
-        op->true_value.accept(this);
-        Monotonic ra = result;
-        op->false_value.accept(this);
-        Monotonic rb = result;
-        Monotonic unified = unify(ra, rb);
-
-        if (rcond == Monotonic::Constant) {
-            result = unified;
-            return;
-        }
-
-        bool true_value_ge_false_value = can_prove(op->true_value >= op->false_value);
-        bool true_value_le_false_value = can_prove(op->true_value <= op->false_value);
-
-        bool switches_from_true_to_false = rcond == Monotonic::Decreasing;
-        bool switches_from_false_to_true = rcond == Monotonic::Increasing;
-
-        if (true_value_ge_false_value &&
-            true_value_le_false_value) {
-            // The true value equals the false value.
-            result = ra;
-        } else if ((unified == Monotonic::Increasing || unified == Monotonic::Constant) &&
-                   ((switches_from_false_to_true && true_value_ge_false_value) ||
-                    (switches_from_true_to_false && true_value_le_false_value))) {
-            // Both paths increase, and the condition makes it switch
-            // from the lesser path to the greater path.
-            result = Monotonic::Increasing;
-        } else if ((unified == Monotonic::Decreasing || unified == Monotonic::Constant) &&
-                   ((switches_from_false_to_true && true_value_le_false_value) ||
-                    (switches_from_true_to_false && true_value_ge_false_value))) {
-            // Both paths decrease, and the condition makes it switch
-            // from the greater path to the lesser path.
-            result = Monotonic::Decreasing;
+        // The result is the unified bounds, added to the "bump" that happens when switching from true to false.
+        if (op->type.is_scalar()) {
+            op->condition.accept(this);
+            ConstantInterval rcond = result;
+
+            op->true_value.accept(this);
+            ConstantInterval ra = result;
+            op->false_value.accept(this);
+            ConstantInterval rb = result;
+            ConstantInterval unified = unify(ra, rb);
+
+            // TODO: How to handle unsigned values?
+            Expr delta = simplify(op->true_value - op->false_value);
+            delta.accept(this);
+            ConstantInterval rdelta = result;
+
+            ConstantInterval adjusted_delta;
+            if (const int64_t *const_delta = as_const_int(delta)) {
+                adjusted_delta = multiply(rcond, *const_delta);
+            } else {
+                adjusted_delta = multiply(rcond, rdelta);
```

Comment:
I think this is wrong. An example where it produces an incorrect result is: `select(0 < x, max(min(x, 4), 3), 4)`

I was hoping to avoid bringing back the old complicated logic... but I haven't yet figured out a cleaner way to handle this.

Reply:
I think we don't want rdelta here, I think we actually want bounds_of_expr(delta) instead. i.e. it should be the actual delta, not the derivative of the delta.

Reply:
I never reviewed this part because I didn't really understand it all that much, but I would like to understand better. Could you explain the comment "// The result is the unified bounds, added to the "bump" that happens when switching from true to false." ? Why is the bump added, and not just a unification like with min and max?

Reply:
My reasoning is: the bounds of the finite difference is the union of the bounds of the finite difference of the true value and the false value, plus the step function that occurs when switching from false to true.